packages/flutter_tools/lib/src/commands/daemon.dart

```diff
@@ -505,7 +505,7 @@ class AppDomain extends Domain {
   }
 }
 
-typedef void _DeviceEventHandler(Device device);
+typedef Future<Null> _DeviceEventHandler(Device device);
```

Comment:
I guess the issue is awaiting `_deviceToMap()`.  The listener callbacks will be put in the microtask queue in order, but will their corresponding awaits of `_deviceToMap()` return in the same order?